be/src/vec/sink/writer/async_result_writer.cpp

```diff
@@ -130,42 +138,47 @@ void AsyncResultWriter::process_block(RuntimeState* state, RuntimeProfile* profi
     }
 
     DCHECK(_dependency);
-    if (_writer_status.ok()) {
-        while (true) {
-            ThreadCpuStopWatch cpu_time_stop_watch;
-            cpu_time_stop_watch.start();
-            Defer defer {[&]() {
-                if (state && state->get_query_ctx()) {
-                    state->get_query_ctx()->resource_ctx()->cpu_context()->update_cpu_cost_ms(
-                            cpu_time_stop_watch.elapsed_time());
-                }
-            }};
-            if (!_eos && _data_queue.empty() && _writer_status.ok()) {
-                std::unique_lock l(_m);
-                while (!_eos && _data_queue.empty() && _writer_status.ok()) {
-                    // Add 1s to check to avoid lost signal
-                    _cv.wait_for(l, std::chrono::seconds(1));
-                }
+    while (_writer_status.ok()) {
+        ThreadCpuStopWatch cpu_time_stop_watch;
+        cpu_time_stop_watch.start();
+        Defer defer {[&]() {
+            if (state && state->get_query_ctx()) {
+                state->get_query_ctx()->resource_ctx()->cpu_context()->update_cpu_cost_ms(
+                        cpu_time_stop_watch.elapsed_time());
             }
+        }};
 
-            if ((_eos && _data_queue.empty()) || !_writer_status.ok()) {
-                _data_queue.clear();
-                break;
+        //1) wait scan operator write data
+        if (!_eos && _data_queue.empty() && _writer_status.ok()) {
```

Comment:
```suggestion
        {
            std::unique_lock l(_m);
            while (!_eos && _data_queue.empty() && _writer_status.ok()) {
                // Add 1s to check to avoid lost signal
                _cv.wait_for(l, std::chrono::seconds(1));
            }
        }
```
I think we can solve this problem by simply deleting this if condition.  The issue of out-of-order execution will be synchronized through the lock later.

Reply:
The if condition is removed. At the same time, for the rigor and readability of the code, the data is taken from the queue or moved down.